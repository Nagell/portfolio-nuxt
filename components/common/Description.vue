<template>
    <CommonTypography
        v-if="type === 'ul'"
        tag="ul"
        class="my-0 text-muted-foreground"
    >
        <CommonTypography
            v-for="(point, key) in parsedPoints"
            :key="key"
            tag="li"
            class="text-pretty"
        >
            <CommonMarkdownText :content="point" />
        </CommonTypography>
    </CommonTypography>
    <div v-if="type === 'p'">
        <CommonTypography
            v-for="(point, key) in parsedPoints"
            :key="key"
            tag="p"
            class="text-pretty"
        >
            <CommonMarkdownText :content="point" />
        </CommonTypography>
    </div>
</template>

<script lang="ts" setup>
    interface Props {
        description: string
        type?: 'ul' | 'p'
    }

    const props = withDefaults(defineProps<Props>(), {
        type: 'ul'
    })

    /** Sanitize and split the description into an array */
    const parsedPoints = computed(() => {
        return props.description.split('\n')
            .filter(item => item !== '')
            .map(item => item.trim())
    })
</script>
