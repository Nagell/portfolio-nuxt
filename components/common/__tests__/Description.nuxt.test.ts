import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'

import Description from '../Description.vue'

const multilineText = `Test description
with multiple lines
of text`

const singleLineText = 'Test description with multiple lines of text'

describe('Description', async () => {
    const wrapper = await mountSuspended(Description, {
        props: {
            description: multilineText,
        },
    })

    describe('default', () => {
        it('renders correctly', async () => {
            expect(wrapper.html()).toMatchSnapshot()
        })

        it('text is split into multiple lines in "li" tags', async () => {
            expect(wrapper.html()).not.toContain(singleLineText)
            expect(wrapper.findAll('ul').length).toBe(1)
            expect(wrapper.findAll('li').length).toBe(3)
        })
    })
})

describe('Description', async () => {
    const wrapper = await mountSuspended(Description, {
        props: {
            description: multilineText,
            type: 'p'
        },
    })

    describe('p', () => {
        it('renders correctly', async () => {
            expect(wrapper.html()).toMatchSnapshot()
        })

        it('text is split into multiple lines in "p" tags', async () => {
            expect(wrapper.html()).not.toContain(singleLineText)
            expect(wrapper.findAll('p').length).toBe(3)
        })
    })
})
