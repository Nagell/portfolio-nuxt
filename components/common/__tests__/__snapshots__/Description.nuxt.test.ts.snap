// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Description Component > default > renders correctly 1`] = `
"<ul class="ml-6 list-disc [&amp;>li]:mt-2 my-0 text-muted-foreground">
  <li class="leading-6 text-pretty">Test description</li>
  <li class="leading-6 text-pretty">with multiple lines</li>
  <li class="leading-6 text-pretty">of text</li>
</ul>
<!--v-if-->"
`;

exports[`Description Component > p > renders correctly 1`] = `
"<!--v-if-->
<div>
  <p class="leading-6 [&amp;:not(:first-child)]:mt-6 text-pretty">Test description</p>
  <p class="leading-6 [&amp;:not(:first-child)]:mt-6 text-pretty">with multiple lines</p>
  <p class="leading-6 [&amp;:not(:first-child)]:mt-6 text-pretty">of text</p>
</div>"
`;
